### PR TITLE
Don't treat object properties as potential JS contructors without JSDoc class tag

### LINKED ISSUE #48915: IntelliSense/autocomplete doesn't work properly inside object literals
Does this issue occur when all extensions are disabled?: Yes

- VS Code Version: 1.66.2 (system setup)
- OS Version: Microsoft Windows Version 21H1 (OS Build 19043.1526)

Steps to Reproduce:

1. Create a new object using object literal with two or more properties and a method:
`let obj = { x: 10, y: [1], fun: function () {}`
2. Now, inside the method, addressing to the properties using `this` keyword is recognised by the IntelliSense/autocomplete.
`this.` - results in
step2
3. Change the value of a property inside the method, for instance
`let obj = { x: 10, y: [1], fun: function () {`**`this.x = 5;`**`}}`
4. Now, only addressing to the property `x` is recognised by the IntelliSense/autocomplete:
step4
    Even though `y` property is an array, IntelliSense won't show its methods and properties:
step4 2

<details><summary>System info</summary>
<table>
  <tr>
    <td>CPUs</td>
    <td>Intel(R) Core(TM) i5-6600 CPU @ 3.30GHz (4 x 3312)</td>
  </tr>
  <tr>
    <td>GPU Status</td>
    <td>2d_canvas: enabled canvas_oop_rasterization: disabled_off direct_rendering_display_compositor: disabled_off_ok gpu_compositing: enabled multiple_raster_threads: enabled_on oop_rasterization: enabled opengl: enabled_on rasterization: enabled raw_draw: disabled_off_ok skia_renderer: enabled_on video_decode: enabled video_encode: enabled vulkan: disabled_off webgl: enabled webgl2: enabled</td>
  </tr>
  <tr>
    <td>Load (avg)</td>
    <td></td>
  </tr>
  <tr>
    <td>Memory (System)</td>
    <td>15.89GB (4.73GB free)</td>
  </tr>
  <tr>
    <td>Process Argv</td>
    <td>...\vscode issue --crash-reporter-id 85c4d140-e4d5-49f1-8a5a-495612b0049f</td>
  </tr>
  <tr>
    <td>Screen Reader</td>
    <td>no</td>
  </tr>
  <tr>
    <td>VM</td>
    <td>0%</td>
  </tr>
</table>
</details>
Extensions: none
<br></br>
<details><summary>A/B experiment info</summary>
<pre>
vsliv368cf:30146710
vsreu685:30147344
python383:30185418
vspor879:30202332
vspor708:30202333
vspor363:30204092
pythonvspyl392:30443607
pythontb:30283811
pythonvspyt551:30345470
pythonptprofiler:30281270
vshan820:30294714
vstes263:30335439
vscorecescf:30445987
pythondataviewer:30285071
vscod805cf:30301675
pythonvspyt200:30340761
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
vsaa593:30376534
vsc1dst:30438360
pythonvs932:30410667
wslgetstarted:30449410
pythonvsnew555:30457759
vscscmwlcmt:30465135
cppdebugcf:30475217
</pre>
</details>

### PR DESCRIPTION
**Pull request by jakebailey**
Fixes #48915

This is probably breaking, so for 4.9 if we want it.

### CODE REVIEW

File: tests/baselines/reference/functionExpressionNames.types [hunk 1]
```diff
@@ -31,7 +31,7 @@ var o = {
         this.c = 'nested object'
 >this.c = 'nested object' : "nested object"
 >this.c : any
->this : this
+>this : { C: typeof C; }
```
Comment (jakebailey) [inline tests/baselines/reference/functionExpressionNames.types:34]: This test is explicitly the case that is changing:

```js
var o = {
    C: function () {
        this.c = 'nested object'
    }
}
var og = new o.C();
```

So, I'm not sure if that's good or bad.

**Review by andrewbranch — CHANGES_REQUESTED** [review]
While I can see in the existing code that this probably already works, there doesn’t seem to be a test (at least in ones you added or changed) that uses `/** @class */` inside an object.

Come to think of it, what should the behavior of

```js
const o = {
  SomeClass() {
    this.x = 0;
  }
}
```

be (with and without `@class`)? I don’t know, but let’s add it to tests too so we have a documented position.

**Comment by jakebailey** [comment]
I think that https://github.com/microsoft/TypeScript/pull/49735/files#diff-dbf882263a483218df32c53af90df429250234fb325ba71f8dd34bcbe165fe54 is that case, but you're right that I forgot one that's tagged (I don't know if there is one; I would assume so but I will check).

**Comment by jakebailey** [comment]
I missed that you had a method declaration there and not a function expression. The runtime behavior is the same, so I'm inclined to make it the same. I'll add an explicit test.

**Comment by jakebailey** [comment]
Handled that case. I'm a little bit unsure what the behavior _should_ be for some of those things that changed. #27561 is relevant.

File: tests/baselines/reference/objectPropertyAsClass.types [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+>{    /** @class */    foo() {        this.x = 0;    }} : { foo(): void; }
+
+    /** @class */
+    foo() {
+>foo : typeof b1.foo
+
+        this.x = 0;
+>this.x = 0 : 0
+>this.x : any
+>this : { foo(): void; }
```
Comment (jakebailey) [inline tests/baselines/reference/objectPropertyAsClass.types:46]: This should say `this : this`, so I screwed something up.

**Comment by andrewbranch** [comment]
I’m not sure I think it’s very important to make object literal method declarations classable. That code looks bananas to me.

File: tests/baselines/reference/objectPropertyAsClass.types (hunk 2 above)
Comment (jakebailey) [reply tests/baselines/reference/objectPropertyAsClass.types:46]: Per the PR comments, we're not going to do `@class` and such for property methods, so, resolving.

**Comment by jakebailey** [comment]
In https://github.com/microsoft/TypeScript/issues/48915#issuecomment-1170278046, I think so, yes. I don't know if this is urgent enough to go into 4.8.

**Comment by andrewbranch** [comment]
I think the breakiness is fairly low, but the bug is not a recent regression, has no upvotes, and can be worked around, so for me it doesn’t meet the bar for considering it now.

**Pull request merged**